### PR TITLE
Add `Config.val_json_bytes`

### PR DESCRIPTION
**Pull request by josh-newman**
## Change Summary

Update `Config` with new option added in pydantic/pydantic-core#1308.

## Related issue number

None.

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: tests/test_json.py [hunk 1]
```diff
@@ -500,3 +501,45 @@ class A(BaseModel):
     m = A(a=MyEnum.A, b=[1, 2, 3], c=Decimal('0'))
     assert m.model_dump_json() == '{"a":"A","b":"list!","c":"decimal!"}'
     assert m.model_dump() == {'a': MyEnum.A, 'b': [1, 2, 3], 'c': Decimal('0')}
+
+
+# Expect failure until pydantic-core is upgraded to include pydantic/pydantic-core#1308.
+@pytest.mark.xfail
```
Comment (sydney-runkle) [inline tests/test_json.py:507]:
```suggestion
@pytest.mark.xfail(reason='<something about pydantic-core upgrade>')
```

**Review by sydney-runkle — COMMENTED** [review]
Looks great! You can use a specific `xfail` reason argument instead of a comment above :).

We can go ahead and merge this before the corresponding `pydantic-core` release is out, but could you please provide a screenshot that these tests are passing locally with your branch of `pydantic-core`?

**Comment by sydney-runkle** [comment]
Let's add the changes in https://github.com/pydantic/pydantic/pull/9772 to this PR as well :). Thanks!

**Comment by josh-newman** [comment]
Thanks @sydney-runkle, I've updated the xfail reasons and run tests at pydantic/pydantic-core#1308's HEAD:
Screenshot 2024-07-05 at 17 53 35

When pydantic/pydantic-core#1308 is incorporated, cherry-picking d7b2fd292166fe67b547cd2ca83cc55818eeb15b should fix the tests.

**Comment by sydney-runkle** [comment]
Just provided some feedback on https://github.com/pydantic/pydantic-core/pull/1308#pullrequestreview-2186661660, thanks for your work here @josh-newman!

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Awesome work! Just a quick note on the docs

**Comment by sydney-runkle** [comment]
Awesome job on getting the tests to pass locally.

> We can go ahead and merge this before the corresponding pydantic-core release is out, but could you please provide a screenshot that these tests are passing locally with your branch of pydantic-core?

I realize, I want to walk this back just a bit - given that this is a new feature and not a bug fix, we'll want to wait for a release. That being said, I can do a minor release of `pydantic-core` once we get your core PR in (should be quite soon), then immediately merge this one as well.

Sorry for the delay on my review, was out for a long holiday. We're super excited about this new feature and your great work!

**Comment by sydney-runkle** [comment]
^^ That being said, I'm not worried about the integration tests failing on the `pydantic-core` end, we can merge that despite the failures knowing that things are working end to end locally.

**Comment by sydney-runkle** [comment]
> I realize, I want to walk this back just a bit - given that this is a new feature and not a bug fix, we'll want to wait for a release. That being said, I can do a minor release of pydantic-core once we get your core PR in (should be quite soon), then immediately merge this one as well.

To elaborate a bit more, in case others are curious in the future: merging this into main would result in new docs being added to our `latest` page that were a bit inaccurate, as said features wouldn't yet be available. Thus, I think it makes more sense to wait. It's a bit of an awkward dance with things not in one repo, but we definitely appreciate having the PRs open in both `pydantic` and `pydantic-core` so that we can see what upstream changes look like.

File: docs/errors/validation_errors.md [hunk 2]
```diff
@@ -93,6 +93,26 @@ except ValidationError as exc:
     #> 'bool_type'
 ```
 
+## `bytes_invalid_encoding`
+
+This error is raised when a `bytes` value is valid under the configured encoding:
```
[the remark this answers is not in the document]
Comment (josh-newman) [reply docs/errors/validation_errors.md]: Thanks, done!

**Review by sydney-runkle — APPROVED** [review]
Looks good, will wait to merge until we pull the `pydantic-core` changes in 👍

**Comment by sydney-runkle** [comment]
Awesome work! Merged into main 🥳 !

**Comment by josh-newman** [comment]
Thanks for the reviews!

**Pull request merged**